kratos/solving_strategies/convergencecriterias/residual_criteria.h

```diff
@@ -325,18 +353,35 @@ class ResidualCriteria
         TDataType residual_solution_norm = TDataType();
         SizeType dof_num = 0;
 
-        // Loop over Dofs
-        #pragma omp parallel for reduction(+:residual_solution_norm,dof_num)
-        for (int i = 0; i < static_cast<int>(rDofSet.size()); i++) {
-            auto it_dof = rDofSet.begin() + i;
-
-            IndexType dof_id;
-            TDataType residual_dof_value;
+        // Auxiliar values
+        TDataType residual_dof_value = 0.0;
+        const auto it_dof_begin = rDofSet.begin();
+        const int number_of_dof = static_cast<int>(rDofSet.size());
 
-            if (it_dof->IsFree()) {
-                dof_id = it_dof->EquationId();
-                residual_dof_value = TSparseSpace::GetValue(b,dof_id);
-                residual_solution_norm += std::pow(residual_dof_value, 2);
+        // Loop over Dofs
+        if (rModelPart.NumberOfMasterSlaveConstraints() > 0) {
+            #pragma omp parallel for firstprivate(residual_dof_value) reduction(+:residual_solution_norm, dof_num)
+            for (int i = 0; i < number_of_dof; i++) {
+                auto it_dof = it_dof_begin + i;
+
+                const IndexType dof_id = it_dof->EquationId();
+
+                if (mActiveDofs[dof_id]) {
+                    residual_dof_value = TSparseSpace::GetValue(b,dof_id);
+                    residual_solution_norm += std::pow(residual_dof_value, 2);
+                    dof_num++;
+                }
+            }
+        } else {
+            #pragma omp parallel for firstprivate(residual_dof_value) reduction(+:residual_solution_norm, dof_num)
+            for (int i = 0; i < number_of_dof; i++) {
+                auto it_dof = it_dof_begin + i;
+
+                if (!it_dof->IsFixed()) {
+                    const IndexType dof_id = it_dof->EquationId();
+                    residual_dof_value = TSparseSpace::GetValue(b,dof_id);
+                    residual_solution_norm += std::pow(residual_dof_value, 2);
+                }
                 dof_num++;
```

Comment:
shouldn't this also be in the loop?

Reply:
you are probably right, although i do not completely understand the use of it

Reply:
used here:
https://github.com/KratosMultiphysics/Kratos/blob/1ca3a824725fa4a727ce649bbb3bf7eb76678760/kratos/solving_strategies/convergencecriterias/residual_criteria.h#L177-L178

If it is outside the loop then it is incorrect

Reply:
fixed it